### PR TITLE
Adds fft namespace

### PR DESCRIPTION
**Pull request by mruberry**
This PR creates a new namespace, torch.fft (torch::fft) and puts a single function, fft, in it. This function is analogous to is a simplified version of NumPy's [numpy.fft.fft](https://numpy.org/doc/1.18/reference/generated/numpy.fft.fft.html?highlight=fft#numpy.fft.fft) that accepts no optional arguments. It is intended to demonstrate how to add and document functions in the namespace, and is not intended to deprecate the existing torch.fft function.

Adding this namespace was complicated by the existence of the torch.fft function in Python. Creating a torch.fft Python module makes this name ambiguous: does it refer to a function or module? If the JIT didn't exist, a solution to this problem would have been to make torch.fft refer to a callable class that mimicked both the function and module. The JIT, however, cannot understand this pattern. As a workaround it's required to explicitly `import torch.fft` to access the torch.fft.fft function in Python:

```
import torch.fft

t = torch.randn(128, dtype=torch.cdouble)
torch.fft.fft(t)
```

See https://github.com/pytorch/pytorch/issues/42175 for future work. Another possible future PR is to get the JIT to understand torch.fft as a callable class so it need not be imported explicitly to be used.

### CODE REVIEW

File: docs/source/fft.rst [hunk 1]
```diff
@@ -0,0 +1,21 @@
+.. role:: hidden
+    :class: hidden-section
+
+torch.fft
+===================================
+
+Discrete Fourier transforms and related functions.
+
+To use these functions the torch.fft module must be imported since its name
+conflicts with the :func:`torch.fft` function.
```
Comment (peterbell10) [inline docs/source/fft.rst:10]: I'm not necessarily recommending this as it's definitely a hack, but you can workaround this issue. Change the `fft` module's `__class__` with a class that has a static `__call__` method defined. We used this when adding `scipy.fft`:
https://github.com/scipy/scipy/blob/4e3d82268eae3c89e94ff5f3c3774bb12c12982a/scipy/fft/__init__.py#L101-L115

Comment (mruberry) [reply docs/source/fft.rst:10]: Hey @peterbell10! That's a good idea and I tried it but it's not currently jittable. I'm following up with the JIT team on the possibility of making that work with the jit, but I don't expect that to be a reasonable solution.

Comment (mruberry) [reply docs/source/fft.rst:10]: (Also, how much of a hack is it when Guido himself tells us to do it?)

Comment (peterbell10) [reply docs/source/fft.rst:10]: It breaks things like sphinx autodoc which try to do inspection on module which isn't great. I certainly wouldn't design an API like that intentionally.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1437,6 +1437,13 @@
     CPU, CUDA: native_group_norm_backward
 
 # FFT
+- func: fft.fft(Tensor self) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml:1440]: what's the plan for the this schema?  It will differ from fft below?  I guess we are okay with mobile because I doubt anyone uses fft on mobile.

Also, we should probably allow namespaces eventually which would make this fft::fft.fft which is a little weird, but whatever.

File: docs/source/fft.rst (hunk 1 above)
Comment (gchanan) [inline docs/source/fft.rst:10]: nit: I'd say something like the deprecated, soon to be deprecated, etc.

File: test/test_spectral_ops.py [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+        def method_fn(t):
+            return t.fft(1)
+        scripted_method_fn = torch.jit.script(method_fn)
+
+        self.assertEqual(scripted_method_fn(t), eager_result)
+
+        with self.assertRaisesRegex(TypeError, "'module' object is not callable"):
+            torch.fft(t, 1)
+
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
```
Comment (gchanan) [inline test/test_spectral_ops.py:41]: it looks like you are moving and changing in one PR (i.e. I see torch.fft.fft which doesn't exist currently).  I don't think we should be doing that -- it seems totally fair to move everything in a previous PR.

File: tools/autograd/gen_python_functions.py [hunk 4]
```diff
@@ -165,6 +191,7 @@ def get_py_torch_functions(declarations):
     def should_bind(declaration):
         return (should_generate_python_binding(declaration) and
                 not is_nn_module_function(declaration) and
+                not is_fft_module_function(declaration) and
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py:194]: err...should we change this around so it's no namespace?  (Might need to leave nn module because of THNN stuff)

File: torch/csrc/autograd/python_fft_functions.h [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <Python.h>
+#include <pybind11/pybind11.h>
```
Comment (gchanan) [inline torch/csrc/autograd/python_fft_functions.h:4]: we do this for nn?  We should limit our header includes to what's needed by the header, but okay.

File: torch/__init__.py [hunk 6]
```diff
@@ -190,7 +190,7 @@ def _load_global_deps():
 
 # Appease the type checker; ordinarily this binding is inserted by the
 # torch._C module initialization code in C
-if False:
+if not _C:
```
Comment (gchanan) [inline torch/__init__.py:193]: why this change?

**Review by gchanan — COMMENTED** [review]
what's the level of JIT support?

File: torch/csrc/autograd/python_fft_functions.h (hunk 5 above)
Comment (mruberry) [reply torch/csrc/autograd/python_fft_functions.h:4]: We do: 

https://github.com/pytorch/pytorch/blob/3c6fae6567230da0e08788b968cfb466ec629d65/torch/csrc/autograd/python_nn_functions.h#L4

I agree with you but it seems like a separate issue.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml]: This is temporarily addressed with a naming pattern and manual Python and C++ bindings, analogous to what nn does.

File: docs/source/fft.rst (hunk 1 above)
Comment (mruberry) [reply docs/source/fft.rst:10]: We may be able to do something like this in the future after some JIT work.

Comment (mruberry) [reply docs/source/fft.rst:10]: It'll have to be in a follow-up since the torch.fft functionality is not the same as numpy.fft.fft's, so we'll need a more thorough deprecation plan.

File: tools/autograd/gen_python_functions.py (hunk 4 above)
Comment (mruberry) [reply tools/autograd/gen_python_functions.py:194]: Let's leave this for the torch.linalg namespace PR?

File: torch/__init__.py (hunk 6 above)
Comment (mruberry) [reply torch/__init__.py:193]: It appeases more type checkers but isn't necessary. I reverted this changed.

File: test/test_spectral_ops.py (hunk 3 above)
Comment (mruberry) [reply test/test_spectral_ops.py:41]: This was addressed with https://github.com/pytorch/pytorch/pull/42157.

File: aten/src/ATen/native/native_functions.yaml [hunk 7]
```diff
@@ -1479,6 +1479,13 @@
     CPU, CUDA: native_group_norm_backward
 
 # FFT
+# NOTE: functions in the fft python_module should have their names start with
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:1482]: Name this note

File: torch/csrc/api/include/torch/fft.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+/// Example:
+/// ```
+/// auto t = torch::randn(128, dtype=kComplexDouble);
+/// torch::fft::fft(t);
+/// ```
+inline Tensor fft(const Tensor& self) {
+  return detail::fft(self);
+}
+
+}} // torch::fft
```
Comment (mruberry) [inline torch/csrc/api/include/torch/fft.h:30]: newline

File: torch/csrc/api/include/torch/fft.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+namespace detail {
+
+inline Tensor fft(const Tensor& self) {
+  return torch::fft_fft(self);
+}
+
+} // namespace detail
+#endif /* DOXYGEN_SHOULD_SKIP_THIS */
+
+/// See the documenation of torch.fft.fft.
```
Comment (mruberry) [inline torch/csrc/api/include/torch/fft.h:19]: documenation

File: test/cpp/api/fft.cpp [hunk 10]
```diff
@@ -0,0 +1,28 @@
+#include <gtest/gtest.h>
+
+#include <torch/torch.h>
+#include <torch/fft.h>
+
+#include <test/cpp/api/support.h>
+
+// NOTE: Visual Studio doesn't understand complex literals
+TEST(FFTTest, fft) {
```
Comment (mruberry) [inline test/cpp/api/fft.cpp:9]: Remove this

File: torch/csrc/api/include/torch/fft.h [hunk 11]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <torch/arg.h>
```
Comment (mruberry) [inline torch/csrc/api/include/torch/fft.h:3]: Include something sensible

File: torch/csrc/autograd/python_fft_functions.h (hunk 5 above)
Comment (mruberry) [reply torch/csrc/autograd/python_fft_functions.h:4]: Note from review: trying building without these headers here.

Comment (mruberry) [reply torch/csrc/autograd/python_fft_functions.h:4]: Try forward declaring the PyObject instead of including Python.h.

**Review by gchanan — APPROVED** [review]
